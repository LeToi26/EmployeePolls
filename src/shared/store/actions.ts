import { InitialState, Questions, User, Users } from "shared/model/poll";
import productApi from "../api/productApi";
import { Dispatch } from "@reduxjs/toolkit";

export const RECEIVE_USERS = "RECEIVE_USERS";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const LOGOUT_CURRENT_USER = "LOGOUT_CURRENT_USER";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";

export function initialData() {
  return (dispatch: Dispatch) => {
    return productApi
      .getInitialData()
      .then(({ users, questions }: { users: any; questions: any }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      });
  };
}

export function receiveUsers(users: Users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function receiveQuestions(questions: Questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleLogout() {
  return (dispatch: Dispatch) => {
    return dispatch(logoutCurrentUser());
  };
}

export function logoutCurrentUser() {
  return {
    type: LOGOUT_CURRENT_USER,
  };
}

export function handleLogin(username: string, password: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    const user: any = Object.values(State.users).find(
      (user: any) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setCurrentUser(user));
    }
  };
}

export function setCurrentUser(authedUser: User | {}) {
  return {
    type: SET_CURRENT_USER,
    authedUser,
  };
}

export function handleAddAnswer(questionId: string, answer: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    return productApi
      .saveQuestionAnswer(State.authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addAnswerQuestion(State.authedUser.id, questionId, answer));
        dispatch(addAnswerUser(State.authedUser.id, questionId, answer));
      });
  };
}

function addAnswerQuestion(author: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function addAnswerUser(authedUser: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

export function handleCreateQuestion(
  firstOption: string,
  secondOption: string
) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();

    return productApi
      .saveQuestion(firstOption, secondOption, State.authedUser)
      .then((question) => {
        dispatch(createQuestion(question));
        dispatch(createQuestionUser(question));
      });
  };
}

export function createQuestionUser(question: any) {
  const { author, id } = question;
  return {
    type: CREATE_QUESTION_USER,
    author,
    qid: id,
  };
}

function createQuestion(question: any) {
  return {
    type: CREATE_QUESTION,
    question,
  };
}
