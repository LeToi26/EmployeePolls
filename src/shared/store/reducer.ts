import { InitialState } from "shared/model/poll";
import {
  SET_CURRENT_USER,
  LOGOUT_CURRENT_USER,
  RECEIVE_USERS,
  RECEIVE_QUESTIONS,
  CREATE_QUESTION,
  CREATE_QUESTION_USER,
  ADD_ANSWER_USER,
  ADD_ANSWER_QUESTION,
} from "./actions";

const initialState: InitialState = {
  authedUser: null,
  users: {},
  questions: {},
};

export default function State(state = initialState, action: any) {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        authedUser: action.authedUser,
      };
    case LOGOUT_CURRENT_USER:
      return {
        ...state,
        authedUser: null,
      };
    case RECEIVE_USERS:
      return {
        ...state,
        users: action.users,
      };
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        questions: action.questions,
      };
    case ADD_ANSWER_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.authedUser]: {
            ...state.users[action.authedUser as keyof typeof state],
            answers: {
              ...state.users[action.authedUser as keyof typeof state].answers,
              [action.qid]: action.answer,
            },
          },
        },
      };
    case ADD_ANSWER_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.qid]: {
            ...state.questions[action.qid as keyof typeof state],
            [action.answer]: {
              ...state.questions[action.qid as keyof typeof state][
                action.answer
              ],
              votes: state.questions[action.qid as keyof typeof state][
                action.answer
              ].votes.concat(action.author),
            },
          },
        },
      };
    case CREATE_QUESTION_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.author]: {
            ...state.users[action.author as keyof typeof state],
            questions: state.users[
              action.author as keyof typeof state
            ].questions.concat(action.qid),
          },
        },
      };
    case CREATE_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.question.id]: action.question,
        },
      };
    default:
      return state;
  }
}
