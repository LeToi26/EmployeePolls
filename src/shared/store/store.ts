import { configureStore } from "@reduxjs/toolkit";
import State from "./reducer";

export const store = configureStore({
  reducer: {
    State,
  },
});
