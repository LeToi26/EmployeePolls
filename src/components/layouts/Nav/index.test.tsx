/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "shared/store/store";
import { BrowserRouter } from "react-router-dom";
import Nav from "./index";
import { setCurrentUser } from "shared/store/actions";

describe("Nav", () => {
  it("should render the component", () => {
    store.dispatch(setCurrentUser({ id: "sarahedo", password: "password123" }));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});
