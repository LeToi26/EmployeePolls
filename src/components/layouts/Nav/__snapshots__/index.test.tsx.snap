// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="bg-gray-800"
        data-headlessui-state=""
      >
        <div
          class="mx-auto px-2 sm:px-6 lg:px-8"
        >
          <div
            class="relative flex h-16 items-center justify-between"
          >
            <div
              class="absolute inset-y-0 left-0 flex items-center sm:hidden"
            >
              <button
                aria-expanded="false"
                class="inline-flex items-center justify-center rounded-md p-2 text-gray-400 hover:bg-gray-700 hover:text-white focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
                data-headlessui-state=""
                id="headlessui-disclosure-button-:r0:"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  Open main menu
                </span>
                <svg
                  aria-hidden="true"
                  class="block h-6 w-6"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
            </div>
            <div
              class="flex flex-1 items-center justify-center sm:items-stretch sm:justify-start"
            >
              <div
                class="flex flex-shrink-0 items-center"
              >
                <img
                  alt="Your Company"
                  class="block h-8 w-auto lg:hidden"
                  src="https://tailwindui.com/img/logos/mark.svg?color=indigo&shade=500"
                />
                <img
                  alt="Your Company"
                  class="hidden h-8 w-auto lg:block"
                  src="https://tailwindui.com/img/logos/mark.svg?color=indigo&shade=500"
                />
              </div>
              <div
                class="hidden sm:ml-6 sm:block"
              >
                <div
                  class="flex space-x-4"
                >
                  <a
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/"
                  >
                    Home
                  </a>
                  <a
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/leaderboard"
                  >
                    Leaderboard
                  </a>
                  <a
                    class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                    href="/add"
                  >
                    New Poll
                  </a>
                </div>
              </div>
            </div>
            <div
              class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
            >
              <button
                class="rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  View notifications
                </span>
                <svg
                  aria-hidden="true"
                  class="h-6 w-6"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14.857 17.082a23.848 23.848 0 005.454-1.31A8.967 8.967 0 0118 9.75v-.7V9A6 6 0 006 9v.75a8.967 8.967 0 01-2.312 6.022c1.733.64 3.56 1.085 5.455 1.31m5.714 0a24.255 24.255 0 01-5.714 0m5.714 0a3 3 0 11-5.714 0"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
              <div
                class="relative ml-3"
                data-headlessui-state=""
              >
                <div>
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    class="flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                    data-headlessui-state=""
                    id="headlessui-menu-button-:r1:"
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      Open user menu
                    </span>
                    <img
                      alt=""
                      class="h-8 w-8 rounded-full"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="bg-gray-800"
      data-headlessui-state=""
    >
      <div
        class="mx-auto px-2 sm:px-6 lg:px-8"
      >
        <div
          class="relative flex h-16 items-center justify-between"
        >
          <div
            class="absolute inset-y-0 left-0 flex items-center sm:hidden"
          >
            <button
              aria-expanded="false"
              class="inline-flex items-center justify-center rounded-md p-2 text-gray-400 hover:bg-gray-700 hover:text-white focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
              data-headlessui-state=""
              id="headlessui-disclosure-button-:r0:"
              type="button"
            >
              <span
                class="sr-only"
              >
                Open main menu
              </span>
              <svg
                aria-hidden="true"
                class="block h-6 w-6"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </div>
          <div
            class="flex flex-1 items-center justify-center sm:items-stretch sm:justify-start"
          >
            <div
              class="flex flex-shrink-0 items-center"
            >
              <img
                alt="Your Company"
                class="block h-8 w-auto lg:hidden"
                src="https://tailwindui.com/img/logos/mark.svg?color=indigo&shade=500"
              />
              <img
                alt="Your Company"
                class="hidden h-8 w-auto lg:block"
                src="https://tailwindui.com/img/logos/mark.svg?color=indigo&shade=500"
              />
            </div>
            <div
              class="hidden sm:ml-6 sm:block"
            >
              <div
                class="flex space-x-4"
              >
                <a
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/"
                >
                  Home
                </a>
                <a
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
                <a
                  class="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
                  href="/add"
                >
                  New Poll
                </a>
              </div>
            </div>
          </div>
          <div
            class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
          >
            <button
              class="rounded-full bg-gray-800 p-1 text-gray-400 hover:text-white focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
              type="button"
            >
              <span
                class="sr-only"
              >
                View notifications
              </span>
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14.857 17.082a23.848 23.848 0 005.454-1.31A8.967 8.967 0 0118 9.75v-.7V9A6 6 0 006 9v.75a8.967 8.967 0 01-2.312 6.022c1.733.64 3.56 1.085 5.455 1.31m5.714 0a24.255 24.255 0 01-5.714 0m5.714 0a3 3 0 11-5.714 0"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
            <div
              class="relative ml-3"
              data-headlessui-state=""
            >
              <div>
                <button
                  aria-expanded="false"
                  aria-haspopup="menu"
                  class="flex rounded-full bg-gray-800 text-sm focus:outline-none focus:ring-2 focus:ring-white focus:ring-offset-2 focus:ring-offset-gray-800"
                  data-headlessui-state=""
                  id="headlessui-menu-button-:r1:"
                  type="button"
                >
                  <span
                    class="sr-only"
                  >
                    Open user menu
                  </span>
                  <img
                    alt=""
                    class="h-8 w-8 rounded-full"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
